src/server/services/procedures/phone-iot/phone-iot.js

```diff
@@ -156,256 +153,264 @@ PhoneIoT.prototype.getColor = function (red, green, blue) {
+        arguments[3] = checkColor;
+        arguments[4] = textColor;
+        arguments[7] = group;
+        arguments[9] = text;
         return this._passToDevice('addRadioButton', arguments);
     };
     /**
      * Add a custom label to the device.
      * Labels are similar to buttons except they cannot be clicked.
      * @param {string} device name of the device (matches at the end)
-     * @param {string} password current password for the device
      * @param {BoundedNumber<0, 100>} x X position of the top left corner of the button (percentage).
      * @param {BoundedNumber<0, 100>} y Y position of the top left corner of the button (percentage).
      * @param {Number=} textColor Color code of the button text (if any) (defaults to black).
-     * @param {String} id Name of the label
-     * @param {string} text The text to display on the button
-     * @returns {boolean} True if the action is successful, false otherwise.
+     * @param {string} id Name of the label
+     * @param {string=} text The text to display on the button (defaults to empty)
      */
-    PhoneIoT.prototype.addLabel = function (device, password, x, y, textColor=this.getColor(0, 0, 0), id, text) {
-        arguments[4] = textColor;
+    PhoneIoT.prototype.addLabel = function (device, x, y, textColor=this.getColor(0, 0, 0), id, text='') {
+        arguments[3] = textColor;
+        arguments[5] = text;
         return this._passToDevice('addLabel', arguments);
     };
+
+    /**
+     * Checks for authentication, a no-op.
+     * @param {string} device name of the device (matches at the end)
+     */
+    PhoneIoT.prototype.authenticate = async function (device) {
+        return this._passToDevice('authenticate', arguments);
+    };
     /**
      * Begin listening for events such as button presses.
      * @param {string} device name of the device (matches at the end)
-     * @param {string} password current password for the device
-     * @returns {boolean} True if the action is successful, false otherwise.
      */
-    PhoneIoT.prototype.listen = async function (device, password) {
+    PhoneIoT.prototype.listen = async function (device) {
+        this.authenticate(device); // throws on failure - we want this
+        
         const _device = await this._getDevice(device);
-        await _device.authenticate(_device, [password]); // throws on failure - we want this
-
         _device.guiListeners[this.socket.clientId] = this.socket;
-        return true;
+    };
+    /**
+     * Sets the password (login) credentials for this device.
+     * @param {string} device name of the device (matches at the end)
+     * @param {string} password the password to use
+     */
+    PhoneIoT.prototype.setCredentials = async function (device, password) {
+        const _device = await this._getDevice(device);
+        _device.credentials[this.socket.clientId] = password;
```

Comment:
Statefulness like this can be problematic if something happens to the server (ie, it crashes due to another service or something). It seems that it might be better to save these in the database or something.

Also, these should probably expire at some point so it doesn't have unbounded growth.

Reply:
Yeah, another issue with the current way is that if the device loses internet for a few minutes it'll disconnect, wiping the credentials, and later reconnect but now missing cred. So yeah, maybe the db is a good route for that, but in the short term I could add an expiry like you mentioned.